web-app/HeimScripts/heatmap/run.R

```diff
@@ -54,31 +59,85 @@ main <- function(max_rows = 100, sorting = "nodes") {
   list(messages = msgs)
 }
 
-addStats <- function(df, sorting) {
+writeMarkerTable <- function(markerTable){
+  colnames(markerTable) <- c("rowLabel", "biomarker",
+                             "log2FoldChange", "t", "pValue", "adjustedPValue", "B")
+  jsn                   <- toJSON(markerTable, pretty = TRUE)
+  write(jsn, file = markerTableJson)
+}
+
+getMeasurements <- function(df) {
+  if (ncol(df) > 3){
+    return(df[, 3:ncol(df)])
+  }
+  else {
+    return( df[3] )
+  }
+}
+
+
+cleanUpLimmaOutput <- function() {
+  if (file.exists(markerTableJson)) {
+    file.remove(markerTableJson)
+  }
+}
+
+hasTwoSubsets <- function(measurements) {
+  getSubset1Length(measurements) < ncol(measurements)
+}
+
+addStats <- function(df, sorting, ranking, max_rows) {
+  measurements  <- getMeasurements(df)
+  rankingMethod <- getRankingMethod(ranking)
+  twoSubsets    <- hasTwoSubsets(measurements)
   if (ncol(df) > 3) {
     #this is the case for more than 1 sample
     df <-
-      applySorting(df,sorting)  # no need for sorting in one sample case
-    variances <-
-      apply(df[,3:ncol(df)],1,var, na.rm = T)  # Calculating
-    # variance per probe (per row)
-    means <- rowMeans(df[,3:ncol(df)], na.rm = T)  # this is just an
-                                                   # auxiliary column - it will not be
-                                                   # used for JSON.
-    sdses <- apply(df[,3:ncol(df)],1,sd, na.rm = T)  # this is just
-                                                     # an auxiliary column -
-                                                     # it will not be used for JSON.
-    df["MEAN"] <- means
-    df["SD"] <- sdses
-    df["SIGNIFICANCE"] <- variances
-    df <- df[with(df, order(-SIGNIFICANCE)),]
+      applySorting(df,sorting)  # no need for sorting in one sample case, we do it here only
+    if ( twoSubsets ){
+      markerTable  <- getDEgenes(df)  # relies on columns being sorted,
+                                      # with subset1 first, this is because of the way
+    } else {                          # design matrix is being constructed
+      cleanUpLimmaOutput()  # In order to prevent displaying the table from previous run.
+    }
+    useLimma <- !is.function(rankingMethod)  # rankingMethod is either a function or a character "limma"
+    if (useLimma  && !twoSubsets) {  # cannot use any of the limma methods for single subset.
+      stop( paste("Illegal ranking method: ", ranking, " two subsets needed.") )
+    }
+    if (useLimma ) {
+      rankingScore <- markerTable[ranking]
+      if (length(rankingScore) == 1 && is.na(rankingScore[1])) {
+           stop(paste("Illegal ranking method selected: ",ranking))
+        }
+    } else {
+      cleanUpLimmaOutput()  # Delete json file so that it is not displayed
+      rankingScore <-
+        apply(measurements, 1, rankingMethod)  # Calculating
+    }                                          # ranking per probe (per row)
```

Comment:
Ranking methodes can be  var,coVar, normRange, mean, median.
What if there are NAs: how are these handled? in that case for mean, median, var you might want to add argument na.rm=T otherwise the mean/median/var for a row will be NA  if there are one or more missing values. 
similarly, in coVar function sd and mean are used which return NAs as well unless you specify na.rm = T
and in normRange min and max are called, which also return NA if there are NA's unless na.rm=T

(Also think of the case where there are >1 samples ,but there are only 1 items not NA, e.g. if there are 2 samples, and a certain row has one missing value, then for this row a variance and standard deviation cannot be computed (will be NA))